PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/ThreeDRotationEffectFormat.cs

```diff
@@ -0,0 +1,96 @@
+using PowerPointLabs.Utils;
+using Shape = Microsoft.Office.Interop.PowerPoint.Shape;
+using Shapes = Microsoft.Office.Interop.PowerPoint.Shapes;
+using ThreeDFormat = Microsoft.Office.Interop.PowerPoint.ThreeDFormat;
+
+namespace PowerPointLabs.SyncLab.ObjectFormats
+{
+    class ThreeDRotationEffectFormat: Format
+    {
+        private static readonly float TOLERANCE = Single.Epsilon;
+
+        public override bool CanCopy(Shape formatShape)
+        {
+            ThreeDFormat threeD = formatShape.ThreeD;
+            return Math.Abs(threeD.RotationX) > TOLERANCE
+                   || Math.Abs(threeD.RotationY) > TOLERANCE
+                   || Math.Abs(threeD.RotationZ) > TOLERANCE
+                   || threeD.Perspective == MsoTriState.msoTrue
+                   || Math.Abs(threeD.FieldOfView) > TOLERANCE;
+        }
+
+        public override void SyncFormat(Shape formatShape, Shape newShape)
+        {
+            if (!Sync(formatShape, newShape))
+            {
+                Logger.Log(newShape.Type + " unable to sync 3DRotation Format");
+            }
+        }
+
+        public override Bitmap DisplayImage(Shape formatShape)
+        {
+            Bitmap image = GraphicsUtil.ShapeToBitmap(formatShape);
+            return image;
+        }
+
+        private static bool Sync(Shape formatShape, Shape newShape)
+        {
+            ThreeDFormat source = formatShape.ThreeD;
+            ThreeDFormat dest = newShape.ThreeD;
+            
+            try
+            {
+                if (source.PresetThreeDFormat == MsoPresetThreeDFormat.msoPresetThreeDFormatMixed)
+                {
+                    // skip setting PresetCamera if Mixed. setting it throws an error
+                    if (source.PresetCamera != MsoPresetCamera.msoPresetCameraMixed)
+                    {
+                        dest.SetPresetCamera(source.PresetCamera);
+                    }
+                    // set FieldOfView anyway, PresetCamera doesn't seem to set it
+                    dest.FieldOfView = source.FieldOfView;
+
+                    dest.RotationX = source.RotationX;
+                    dest.RotationZ = source.RotationZ;
+                    dest.RotationY = source.RotationY;
+
+                    // set perspective only if it is different,
+                    // it changes the lighting of the shape when set to the same value
+                    // this change cannot be reverted using code
+                    if (dest.Perspective != source.Perspective)
+                    {
+                        dest.Perspective = source.Perspective;
+                    }
+                }
+                else
+                {
+                    dest.SetThreeDFormat(source.PresetThreeDFormat);
+                }
+
+
+                // when source.Z > 4000 or source.Z < - 4000, it actually means 0
+                float nearestZ = source.Z;
+                nearestZ = nearestZ > 4000 ? 0f : nearestZ;
```

Comment:
perhaps you can try to remove magic numbers and place 4000 up above in a constant, and also explain a little on where you get the number from.

Reply:
Good catch there, forgot to mention that Z must be in a range of -4000 < Z < 4000.